schedview/app/prenight/prenight.py

```diff
@@ -1,418 +1,648 @@
+    )
 
-    if nside is None:
-        try:
-            nside = scheduler.nside
-        except AttributeError:
-            nside = observatory.nside
+    timezone = param.ObjectSelector(
+        objects=AVAILABLE_TIMEZONES,
+        default=DEFAULT_TIMEZONE,
+        doc='The timezone in which "local time" is to be shown.',
+        label="Timezone",
+    )
 
-    if isinstance(observations, str):
-        opsim_output_fname = observations
-    else:
-        # If we are passed an array of observations, write them to a file.
-        converter = SchemaConverter()
+    tier = param.ObjectSelector(
+        default="",
+        objects=[""],
+        doc="The label for the first index into the CoreScheduler.survey_lists.",
+        label="Tier",
+    )
+    surveys = param.ListSelector(
+        objects=[],
+        default=[],
+        doc="The labels for the second index into the CoreScheduler.survey_lists.",
+        label="Surveys",
+    )
+    basis_function = param.ObjectSelector(
+        default="",
+        objects=[""],
+        doc="The label for the basis function to be shown.",
+        label="Basis function",
+    )
 
-        # Get a unique temp file name
-        with NamedTemporaryFile(
-            prefix="opsim-", suffix=".db", dir=TEMP_DIR.name
-        ) as temp_file:
-            opsim_output_fname = temp_file.name
+    _observatory = ModelObservatory()
+    _site = _observatory.location
+    # Must declare all of these as Parameters, even though they should not
+    # be set by the user, because they are used in the @depends methods,
+    # and otherwise Parametrized will assume that they depend on
+    # everything.
+    _visits = schedview.param.DataFrame(
+        None,
+        doc="The visits for the night.",
+        columns={
+            "fieldRA": float,
+            "fieldDec": float,
+            "observationStartMJD": float,
+            "filter": str,
+            "rotSkyPos": float,
+            "rotSkyPos_desired": float,
+        },
+    )
+    _scheduler = schedview.param.TypedParameter(
+        None, value_type=rubin_sim.scheduler.schedulers.CoreScheduler
+    )
+    _almanac_events = schedview.param.DataFrame(
+        None,
+        doc="Events from the rubin_sim alamanc",
+        columns={"MJD": float, "LST": float, "UTC": pd.Timestamp},
+    )
 
-        converter.obs2opsim(observations, filename=opsim_output_fname)
+    # To use the convenient panel widgets, we need to use a python stdlib
+    # datetime.date instance to handle the date as provided by the user,
+    # but we often need an instance of astropy.time.Time when we actually
+    # use it. Therefore, use param to convert it on one place, rather than
+    # doing the conversion independently eveywhere we need it.
+    _night_time = schedview.param.TypedParameter(
+        None,
+        value_type=Time,
+        doc="A local time of the start of the night, used to specify which"
+        "night is being examined",
+    )
```

Comment:
so _night_time doesn’t necessarily equal “night” or does it? the fact that it’s described as the start of the night is throwing me off here. is “night” the actual time or the time at the start of the night? and is “night” a time or a (essentially integer) describing which night it is?

Reply:
It is technically a time, but it is used to designate a night, but that's true of all mechanisms for specifying a night that are reasonable in this context.

The rubin_sim approach (numbered night of the survey) doesn't work because we shouldn't be asking observers to keep track of when the survey starts, or make them calculate how far away it is. Similarly, we don't want to put this date on plots because this will then require the user to do the reverse of this to understand what they are looking at. This approach is ambiguous, and leads to multiple different nights with the same night (when the start date of the survey gets rescheduled). For auxtel, it will also be a negative number, which I suppose would work, but would be a bit weird. Ultimately, in the context of dashboards, what we really need is a _calendar date_ because this is what users will want to specify and what they will want to see plotted. (Even in the context of the simulate, I've always found it awkward to deal with, because you need to remember to add the start date to do anything useful with it.)

But, calendar dates have their own problems. The local night "rolls over" at midnight, and the UT night (in Chile) "rolls over" around the start of the night, sometimes before, sometimes after. So, the calendar date "July 11, 2023" is ambiguous: it might mean the night that begins at July 11, 2023 local time, or it might mean the night almost off of which happens during July 11, 2023 UTC (which is a day earlier than the one that begins on July 11, 2023 local time, and so the night that _ends_ on July 11, 2023 local time).

So, to express the night, it needs to be a calendar date (so that a human can specify and interpret it), and it must include a time and time zone. The practice used at the NOAO (now NOIRLab) observatories I have worked with (CTIO and Kitt Peak) is to use the local date of the time of the start of the night. (SDSS used the UT of midnight, so was offset from NOAO by one.)

The panel input widget supports the python stdlib datetime.date widget, which is basically a tuple of year, month, and day. Nothing about the data type, though, indicates whether it should be interpreted as it would be at CTIO (when night that begins when the local date has these values) or as it would have been at SDSS (when UTC has those values at midnight). So, first, it needs to be documented.

Using an astropy.time.Time instance is an unambiguous way to specify a night, for any times we might care about: the night for a given astropy.time.Time instance is the night in which that instant in time occurs. (I specify "we might care about" because the time between local civil and astronomical noon is ambiguous, but we really don't care.) 

Of course, as I write this, I noticed a pseudo-bug it how I converted the datetime.date to and astropy.time.Time. The current way I did it will always give the right answer in Chile, but might be conceptually less clear. I'll fix it, and put the most of the text here into a comment in the code.

Reply:
Thanks - I agree that counting from the start of the survey is awkward and not what we'd want to do, and also that we do need to define what "night" means, so spelling out that it is the calendar date in Chile at sunset is likely useful (and will avoid any confusion if engineering time means that the observing was planned to start after local midnight). 

"night" doesn't change during the operation of this code, unless you're skipping ahead to the next entire night, right?

Reply:
For the pre-night? No. In production, I intend to make the default date the current day, and there usually won't be a reason to change it in the standard use case (preparing for a night of observing). There will be exceptions, though. For example, if a user is in the middle of downtime, and wants to see what will happen on the first night back, they will need to set the date to the first date back, not the current date.
In practice, what the value is used for is to determine which night to report astronomical events for in the table at the top, and what time range to include in the visits in the following plots.